### ISSUE TITLE
Enrich disease list with a few specific llm-generated tags/grouping classes

### ISSUE THREAD

**Issue by piotrkan** [everycure-org/matrix-disease-list#13]
**Context:**  Our medical team found that specific grouping classes (also referred to as tags) are useful when analyzing the disease list downstream - they allow us to focus on specific diseases of interest or analyst the nature of drug-disease pair more effectively. So far our medical team developed and appended such tags for some of the diseases using LLM-based approaches however this is a lot of manual labour that is not versioned and is also a separate feature from the 'raw' disease list. It would be great if such tags became a part of the final disease list.

**Ask:** To include such filters/tags/grouping classes in the disease list so that next release contains them. Some of those tags are relatively simple and might be obtained by simple database calls while others might be challenging to achieve

**Tags of interest**
The tags we are interested in at the moment are as follows (you can think of them as columns in the table): 
- `is_pathogen_caused` - whether a disease is usually caused by a pathogen (bacteria, viruses, parasites, fungi), returns boolean for each disease 
- `is_glucose_dysfunction` - whether a disease is a primarily a disease of glucose dysregulation (eg diabetes), returns boolean
- `is_cancer` - is this disease a cancer, returns boolean

More challenging but also more interesting:
- `tag_existing_treatment` - are there existing efficacious treatments that cure, prevent or treat this disease, returns boolean
- `tag_QALY_lost` - Categorize the quality-adjusted life years lost for disease into specific outputs based on morbidity, mortality, and quality adjusted life years lost.of life; outputs: very high, high, medium, low, none

**How to achieve this**
There is no specific way we want this to be achieved. Simpler tags (e.g. is_cancer) could be potentially achieved with simple database calls however more complex tags (tag_existing_treatment or tag_QALY_lost) might require use of LLMs. This is something @elliottsharp has done and we can happily share the details on how to achieve this using LLMs through API calls.

**Final output** Example of how this final output would look like:
![image](https://github.com/user-attachments/assets/3e043f52-1d13-4ca4-b8a8-400928513df1)

**Comment by matentzn** [comment]
Thank you for your issue @piotrkan.

If I understand this correctly, you are looking for the most natural place to document qualitative features (such as _Disease groupings_ (cancer, pathogen-caused), _Treatment clusters_ (SSRIs, Surgery, Ibuprofen, etc), _Effects on patient_ (High QALY, morbidity, mortality, etc) of a disease moving forward, and believe this place should be the disease list. 

In my view this should be discussed with the Modelling Working Group, but I dont want to block you now - we can implement some solution here and then have people argue about it later. 

My initial sense is that treatment clusters/categories belong with the drug list repo, so lets focus on the other two, disease groupings and effects on patient.

## Disease groupings. 

There are three cases:

1. There already is a suitable disease grouping in the ontology (e.g. for cancer, there is one).
1. There is not a suitable disease grouping, but it makes general sense (e.g. I am making this up now: skeletal system disorder)
1. There is not a suitable disease grouping in Mondo, but it is a very use case-specific grouping (e.g. glucose-dysfunction-related)

To accommodate for the second, I think it makes sense to keep making individual issues here in this repo requesting the addition of a specific grouping, say "glucose-dysfunction-related", and have the Mondo team decide on a case by case basis where it should go. In the meantime, we could create a temporary grouping class as part of the disease list process which reflects that concept. This would at the same time be the solution for the third case.

Once a grouping class exist in the ontology, the "tag" as you say becomes a simple ontology query with that class.

Why adding a grouping class instead of simply hard-coding a tag? Here are my reasons:

1. We can use typical ontology metadata to add documentation about each grouping / tag (editors notes, comments, synonyms etc)
2. The groupings undergo some form of review when they are added, and the review is already an established workflow (i.e. Mondo PRs)
3. Other people in the world can benefit from your groupings - if you find them useful, odds are, others do too.

## Effects on individuals suffering from the disease

Here I am a bit torn. In contrast to the above disease categories, this is venturing deeply into the world of disease features. I am not a great expert on this, but we have people in Monarch that do nothing else. What concerns me is that these appear to me to be dynamic features of a disease, in the sense that they depend greatly on the current state of medical research (every year they might change); moreover, your plan is to extract them with AI-based methods, which by itself is something that should live under the umbrella of your ML-system and undergo some form of evaluation.

I _can be convinced_ to collect this information here in this repo (by the way, you should consider using [ontogpt](https://github.com/monarch-initiative/ontogpt) for this), but I personally feel that it would be be more transparent to use the KG SOP for this (which hopefully exists in some rudimentary manner):

1. Make sure the feature you are trying to add is part of you KG data model
2. Create an ETL ingest that creates instances for that data model (e.g. Disease -> QUALY associations or property assertions)
3. Expert these to KGX and merge them with the KG every time it is rebuilt

If you start adding such dynamic, and valuable, data to places like the disease list you will soon loose overview which data point is generated where..

That said, if this does not convince you:

1. Add a link to a google sheets here to this ticket
2. The sheet contains at least two columns, one with the `mondo_id` and one with the`tag`

I will process them and release them as part of the disease list.

**Comment by matentzn** [comment]
Continuing discussion from https://github.com/everycure-org/matrix/pull/541/files/3930d10eafe053b23bd8038ceae621653d490b0f..5654795923b47f019f7b0a31bd12102b4e05e183#r1815240444

@piotrkan I can see your point now; two follow up questions:

1. The disease list changes all the time, how do you see the tags being updated (the ones requiring AI - some of them can not simply be looked up like the QUALY ones)
2. Lets say we do find a way to maintain the tags moving forward, how do we deal with errors? Like, how can someone go in and change a specific tag if it was wrongly determined by, say, OpenAI GPT-4o, to fix an odd display on the matrix website?

**Comment by piotrkan** [comment]
Thanks for your answers @matentzn! For completeness, replying to the first question here and then the second. 
____
I do think that such improvements should sit within the disease list as it's a public list, meaning that other users of the disease list could benefit from it - just like you said :) . I would not worry about the downstream task performance being affected by the tags as we are not planning to ingest them in the KG or feed them into our model training - you can think about these tags as 'metadata'.

I would be more than happy to have tags which already exist being added through database/ontology querying (in fact, that would be my preferred way of getting the tags). The LLM idea was just a suggestion which worked well for our medical team and could be implemented relatively easily as it's a simple API call. The issue arises for tags for which there are no existing databases (eg QALY) - you are absolutely correct that these disease features might evolve over time and differ between regions and might change every year however with disease list releases, we effectively have a timestamp we can use for tracking. 

I agree that it is a good idea to ingest such tags/features into a knowledge graph and that is something we should do at some point however this is a different scope to idea which can be essentially viewed as expanding the metadata for drugs/diseases
____
Regarding the follow-up questions: for both questions, we will need to have human in the loop; while simple tags can be looked up in a relatively straightforward manner, QUALY tag needs clinician/physician assessment - this is what @elliottsharp together with our medical team has been doing. Doing it for all 16k diseases every time there is an update is completely unreasonable but provided we spend a significant chunk of time curating this disease list in the first place, then every consecutive update should be manageable (even if a clinician/physician needs to review updated 100 diseases) as I suppose those updates won't be super frequent. 

There will inevitably be errors present however these I imagine could also occur not only for tags but also for names/IDs (e.g. let's say that a specific version of MONDO or CHEMBL has an error somewhere in it). There is of course risk of hallucination when it comes to LLMs but this is why such tags would be more suitable at the source where before new disease list is released and ingested, potential differences/errors are being examined by human in the loop.

**Comment by matentzn** [comment]
Alright, I look forward to talking to you in a bit and hash out a solution! What you are saying makes sense!

**Comment by matentzn** [comment]
@marcello-deluca the general feature works now, but we need to implement the _specific_ tags requested by the OP. 

I would suggest to also check the specific prompts @piotrkan was using in his initial work linked above.

**Comment by piotrkan** [comment]
@marcello-deluca @matentzn you can find these prompts with corresponding models/temperatures here https://github.com/everycure-org/matrix/blob/main/pipelines/matrix/conf/base/preprocessing/parameters.yml 

if you want to see how the langchain object is being called, see `preprocessing/tagging.py file`

**Comment by piotrkan** [comment]
@matentzn just to clarify for me - does this mean that the tags will get generated as a part of disease list repo (just like its done for drugs right now)?

**Comment by matentzn** [comment]
Yes, they are - for now! We might eventually change that; but for now they are bundled in the disease list release pipeline.
